### PR TITLE
Game/AI: Implement SMART_TARGET_HOME_POSITION for SMART_ACTION_MOVE_TO_POS

### PR DESCRIPTION
**Pull request by Lumber-Jack-2**
[//]: # (***************************)
[//]: # (** FILL IN THIS TEMPLATE **)
[//]: # (***************************)

**Changes proposed:**

-  This pr should make the script for #18702 more simple.
-  Currently there is no way to check the homeposition by default (we must always set x,y,z,o manually in the sai-script. This PR should make it more simple

**Target branch(es):** 3.3.5/master

- [X] 3.3.5
- [X] master

**Issues addressed:** Closes #  (insert issue tracker number)
- make script for #18702 more simple

**Tests performed:** (Does it build, tested in-game, etc.)
yes and yes

I tested it with the testscript:

```sql
-- William Saldean SAI
SET @ENTRY := 33996;
UPDATE `creature_template` SET `AIName`="SmartAI" WHERE `entry`=@ENTRY;
DELETE FROM `smart_scripts` WHERE `entryorguid`=@ENTRY AND `source_type`=0;
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(@ENTRY,0,0,0,1,0,100,1,6000,6000,0,0,69,1,0,0,5,0,0,19,2118,100,0,0,0,0,0,"William Saldean - Out of Combat - Move To Closest Creature 'Abigail Shiel' (No Repeat)"),
(@ENTRY,0,1,0,1,0,100,0,10000,10000,0,0,69,0,0,0,0,0,0,28,0,0,0,0,0,0,0,"William Saldean - Out of Combat - Move To Home Position");
```

and everything works fine.

Before scriptcall http://i.imgur.com/B64M3qg.jpg, after scriptcall http://i.imgur.com/hbO5c2n.jpg

[//]: # (**DEPRECATION NOTICE** Instead of creating new PRs to the (old) 6.x branch, make sure the target branch is **master** instead of 6.x.)
[//]: # (**NOTE** Enable the setting "[√] Allow edits from maintainers." when creating your pull request (if not enabled already).)
[//]: # (**NOTE** If this PR __only__ contains SQL files, open a new issue instead and post or link the SQL there.)
[//]: # (**SUGGESTION** When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.)

### CODE REVIEW

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 1]
```diff
@@ -1523,6 +1523,9 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
             }
             else
                 me->GetMotionMaster()->MovePoint(e.action.MoveToPos.pointId, target->GetPositionX(), target->GetPositionY(), target->GetPositionZ(), e.action.MoveToPos.disablePathfinding == 0);
+            
+            if (e.GetTargetType() == SMART_TARGET_HOME_POSITION)
+                me->GetMotionMaster()->MoveTargetedHome();
```
Comment (joschiwald) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: you are moving twice if target is null (line 1515)

Comment (Rushor) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: oh yeah, ty. moving it inside if (!target) should be right?

**Comment by joschiwald** [comment]
you are still moving twice

**Comment by Rushor** [comment]
removed

**Comment by joschiwald** [comment]
looks okay

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 2]
```diff
@@ -1519,7 +1519,10 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
                     if (TransportBase* trans = me->GetDirectTransport())
                         trans->CalculatePassengerPosition(dest.x, dest.y, dest.z);
 
-                me->GetMotionMaster()->MovePoint(e.action.MoveToPos.pointId, dest.x, dest.y, dest.z, e.action.MoveToPos.disablePathfinding == 0);
+                if (e.GetTargetType() == SMART_TARGET_HOME_POSITION)
+                    me->GetMotionMaster()->MoveTargetedHome();
```
Comment (joschiwald) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: now, you calculate transport passenger position and dont need it, maybe the best would be a big switch case

Comment (Rushor) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: oh i see, all these checks are so unorganised in sai :(

Comment (Rushor) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: @joschiwald 

like this ?

```c++
        case SMART_ACTION_MOVE_TO_POS:
        {
            if (!IsSmart())
                break;

            WorldObject* target = NULL;

            if (e.GetTargetType() == SMART_TARGET_CREATURE_RANGE || e.GetTargetType() == SMART_TARGET_CREATURE_GUID ||
                e.GetTargetType() == SMART_TARGET_CREATURE_DISTANCE || e.GetTargetType() == SMART_TARGET_GAMEOBJECT_RANGE ||
                e.GetTargetType() == SMART_TARGET_GAMEOBJECT_GUID || e.GetTargetType() == SMART_TARGET_GAMEOBJECT_DISTANCE ||
                e.GetTargetType() == SMART_TARGET_CLOSEST_CREATURE || e.GetTargetType() == SMART_TARGET_CLOSEST_GAMEOBJECT ||
                e.GetTargetType() == SMART_TARGET_OWNER_OR_SUMMONER || e.GetTargetType() == SMART_TARGET_ACTION_INVOKER ||
                e.GetTargetType() == SMART_TARGET_CLOSEST_ENEMY || e.GetTargetType() == SMART_TARGET_CLOSEST_FRIENDLY)
            {
                ObjectList* targets = GetTargets(e, unit);
                if (!targets)
                    break;

                target = targets->front();
                delete targets;
            }

            switch (e.GetTargetType())
            {
                case SMART_TARGET_CREATURE_RANGE:
                case SMART_TARGET_CREATURE_GUID:
                case SMART_TARGET_CREATURE_DISTANCE:
                case SMART_TARGET_GAMEOBJECT_RANGE:
                case SMART_TARGET_GAMEOBJECT_GUID:
                case SMART_TARGET_GAMEOBJECT_DISTANCE:
                case SMART_TARGET_CLOSEST_CREATURE:
                case SMART_TARGET_CLOSEST_GAMEOBJECT:
                case SMART_TARGET_OWNER_OR_SUMMONER:
                case SMART_TARGET_ACTION_INVOKER:
                case SMART_TARGET_CLOSEST_ENEMY:
                case SMART_TARGET_CLOSEST_FRIENDLY:
                    float x, y, z;
                    target->GetPosition(x, y, z);
                    if (e.action.MoveToPos.ContactDistance > 0)
                        target->GetContactPoint(me, x, y, z, e.action.MoveToPos.ContactDistance);
                    me->GetMotionMaster()->MovePoint(e.action.MoveToPos.pointId, x, y, z, e.action.MoveToPos.disablePathfinding == 0);
                    break;
                case SMART_TARGET_HOME_POSITION:
                    me->GetMotionMaster()->MoveTargetedHome();
                    break;
                case SMART_TARGET_POSITION:
                {
                    G3D::Vector3 dest(e.target.x, e.target.y, e.target.z);
                    if (e.action.MoveToPos.transport)
                        if (TransportBase* trans = me->GetDirectTransport())
                            trans->CalculatePassengerPosition(dest.x, dest.y, dest.z);
                    me->GetMotionMaster()->MovePoint(e.action.MoveToPos.pointId, dest.x, dest.y, dest.z, e.action.MoveToPos.disablePathfinding == 0);
                    break;
                }
                default:
                    break;
            }
            break;
        }
```

Comment (Rushor) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: @joschiwald ?

Comment (joschiwald) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: that looks ugly, move dat target stuff inside the switch case

Comment (Rushor) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: i have no idea what you mean, could you post the correct way? it's only 30 lines of code cmon man you can do it

Comment (Rushor) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: any news?

Comment (joschiwald) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]:
```c++
                case SMART_TARGET_CREATURE_RANGE:
                case SMART_TARGET_CREATURE_GUID:
                case SMART_TARGET_CREATURE_DISTANCE:
                case SMART_TARGET_GAMEOBJECT_RANGE:
                case SMART_TARGET_GAMEOBJECT_GUID:
                case SMART_TARGET_GAMEOBJECT_DISTANCE:
                case SMART_TARGET_CLOSEST_CREATURE:
                case SMART_TARGET_CLOSEST_GAMEOBJECT:
                case SMART_TARGET_OWNER_OR_SUMMONER:
                case SMART_TARGET_ACTION_INVOKER:
                case SMART_TARGET_CLOSEST_ENEMY:
                case SMART_TARGET_CLOSEST_FRIENDLY:
                {
                    ObjectList* targets = GetTargets(e, unit);
                    if (!targets)
                        break;

                    WorldObject* target = targets->front();
                    delete targets;

                    float x, y, z;
                    target->GetPosition(x, y, z);
                    if (e.action.MoveToPos.ContactDistance > 0)
                        target->GetContactPoint(me, x, y, z, e.action.MoveToPos.ContactDistance);
                    me->GetMotionMaster()->MovePoint(e.action.MoveToPos.pointId, x, y, z, e.action.MoveToPos.disablePathfinding == 0);
                    break;
                }
```
something like that

**Comment by Rushor** [comment]
SMART_TARGET_POSITION is used wrong here, there need to be a way to get the exact position inside the targettype itself. The whole concept of this pr is wrong

**Pull request closed without merging**